File: entities.py
import pygame
import math
from assets import Assets, Colors

class Entity:
    def __init__(self, x, y, width, height):
        self.rect = pygame.Rect(x, y, width, height)
        # Cinematic Variables
        self.vel_x = 0
        self.vel_y = 0
        self.acc_x = 0
        self.acc_y = 0
        self.angular_vel = 0
        self.angle = 0

    def update_physics(self, dt):
        """
        Integración Cinemática Estricta:
        v = v_0 + a * t
        s = s_0 + v_0 * t + 0.5 * a * t^2 
        """
        # Desplazamiento
        ds_x = self.vel_x * dt + 0.5 * self.acc_x * (dt ** 2)
        ds_y = self.vel_y * dt + 0.5 * self.acc_y * (dt ** 2)
        
        self.rect.x += ds_x
        self.rect.y += ds_y
        
        # Nueva Velocidad
        self.vel_x += self.acc_x * dt
        self.vel_y += self.acc_y * dt
        
        # Rotación
        self.angle += self.angular_vel * dt

    def collides_aabb(self, x, y):
        return self.rect.collidepoint(x, y)

class Button(Entity):
    def __init__(self, x, y, width, height, text, letter=""):
        super().__init__(x, y, width, height)
        self.text = text
        self.letter = letter
        self.is_hovered = False
        self.disabled = False
        self.state_color = Colors.PRIMARY_GREEN # Default green
        self.bg_color = Colors.BG

    def update(self, dt):
        if not self.disabled:
            mouse_pos = pygame.mouse.get_pos()
            self.is_hovered = self.collides_aabb(*mouse_pos)
            self.bg_color = Colors.DARK_GREEN if self.is_hovered else Colors.BG

    def draw(self, surface):
        # Apply transforms if falling (physics)
        if self.angle != 0:
            # We'd use a temporary surface to rotate, but for optimization we just draw rects.
            pass 

        # Draw Base
        pygame.draw.rect(surface, self.bg_color, self.rect)
        pygame.draw.rect(surface, self.state_color, self.rect, 2, border_radius=4)
        
        # Draw text
        font = Assets.get_font(16)
        letter_font = Assets.get_font(20)
        
        text_rect = self.rect.copy()
        
        if self.letter:
            let_surf = letter_font.render(self.letter, True, self.state_color)
            surface.blit(let_surf, (self.rect.x + 10, self.rect.y + 15))
            text_rect.x += 40
            text_rect.width -= 40
        
        text_rect.y += 10
        text_rect.height -= 10
        Assets.render_text_wrapped(surface, self.text, self.state_color, text_rect, font)

class TimerBar:
    def __init__(self, x, y, width, height, total_time):
        self.rect = pygame.Rect(x, y, width, height)
        self.total_time = total_time
        self.time_left = total_time
        self.is_critical = False

    def get_ratio(self):
        return max(0.0, self.time_left / self.total_time)

    def draw(self, surface):
        ratio = self.get_ratio()
        fill_width = int(self.rect.width * ratio)
        fill_rect = pygame.Rect(self.rect.x, self.rect.y, fill_width, self.rect.height)
        
        color = Colors.ALERT_RED if self.is_critical or ratio < 0.25 else Colors.PRIMARY_GREEN
        
        pygame.draw.rect(surface, Colors.DARK_GREEN, self.rect)
        if fill_width > 0:
            pygame.draw.rect(surface, color, fill_rect)
        pygame.draw.rect(surface, Colors.WHITE, self.rect, 1)

class ECGMonitor:
    """ Genera el trazado en vivo del Electrocardiograma usando splines matemáticos procedimentales """
    def __init__(self, x, y, width, height):
        self.rect = pygame.Rect(x, y, width, height)
        self.offset = 0
        self.points = []
        self.color = Colors.PRIMARY_GREEN
        
    def update(self, dt, bpm):
        self.offset += 150 * dt  # Speed of wave
        if self.offset > self.rect.width:
            self.offset = 0
            self.points.clear()
            
        # Generar "onda"
        center_y = self.rect.centery
        local_x = self.offset
        
        # Simplified math wave for heartbeat
        y = center_y
        cycle = (self.offset / (8000/bpm)) % 1.0
        
        if 0.3 < cycle < 0.4:  # QRS Spikes
            sp_phase = (cycle - 0.3) / 0.1
            if sp_phase < 0.33: y += 10
            elif sp_phase < 0.66: y -= 40
            else: y += 20
        elif 0.5 < cycle < 0.6: # T wave
            y -= math.sin((cycle - 0.5) / 0.1 * math.pi) * 15
            
        y += (pygame.time.get_ticks() % 3) - 1 # Noise
        
        self.points.append((self.rect.x + local_x, y))
        
        # Keep points mapped to width boundaries (fade out old)
        self.points = [p for p in self.points if p[0] > self.rect.x]

    def draw(self, surface):
        pygame.draw.rect(surface, (10, 20, 10), self.rect)
        pygame.draw.rect(surface, self.color, self.rect, 1)
        
        if len(self.points) > 1:
            pygame.draw.lines(surface, self.color, False, self.points, 2)

File: entities.js
export class ECGMonitor {
    constructor(canvasId) {
        this.canvas = document.getElementById(canvasId);
        this.ctx = this.canvas.getContext('2d');
        this.width = this.canvas.width;
        this.height = this.canvas.height;
        this.offset = 0;
        this.points = [];
        this.color = '#00ff41'; // Primary Green
    }

    update(dt, bpm, isCodeBlue) {
        this.color = isCodeBlue ? '#ff0055' : '#00ff41';
        this.offset += 150 * dt; // Speed of wave

        if (this.offset > this.width) {
            this.offset = 0;
            this.points = [];
        }

        const centerY = this.height / 2;
        let y = centerY;

        // Cycle calculated based on bpm
        const cycleLength = 8000 / bpm;
        let cycle = (this.offset / cycleLength) % 1.0;

        if (cycle > 0.3 && cycle < 0.4) {
            const sp_phase = (cycle - 0.3) / 0.1;
            if (sp_phase < 0.33) y += 15;
            else if (sp_phase < 0.66) y -= 60;
            else y += 30;
        } else if (cycle > 0.5 && cycle < 0.6) {
            y -= Math.sin(((cycle - 0.5) / 0.1) * Math.PI) * 20;
        }

        // Noise
        y += (Math.random() * 4) - 2;

        this.points.push({ x: this.offset, y: y });
        // Keep points mapped to width boundaries
        this.points = this.points.filter(p => p.x > 0);
    }

    draw() {
        // Fade effect to simulate CRT persistence phosphor
        this.ctx.fillStyle = 'rgba(5, 5, 5, 0.1)';
        this.ctx.fillRect(0, 0, this.width, this.height);

        this.ctx.beginPath();
        this.ctx.strokeStyle = this.color;
        this.ctx.lineWidth = 3;
        this.ctx.shadowBlur = 10;
        this.ctx.shadowColor = this.color;

        for (let i = 0; i < this.points.length; i++) {
            const p = this.points[i];
            if (i === 0) {
                this.ctx.moveTo(p.x, p.y);
            } else {
                this.ctx.lineTo(p.x, p.y);
            }
        }
        this.ctx.stroke();
    }
}

export class KinematicEntity {
    constructor(domElement) {
        this.el = domElement;
        this.velX = 0;
        this.velY = 0;
        this.accX = 0;
        this.accY = 0;
        this.angularVel = 0;
        this.angle = 0;

        // Original positions
        const rect = this.el.getBoundingClientRect();
        this.x = 0; // We will use CSS transforms relative to their original steady state
        this.y = 0;

        // Save inline styles if necessary
        this.isActive = false;
    }

    triggerExplosion() {
        this.isActive = true;
        this.accY = 800 + Math.random() * 700; // Gravity
        this.velX = (Math.random() - 0.5) * 400; // Explosion horizontal
        this.angularVel = (Math.random() - 0.5) * 600; // Rotation speed
        this.el.style.position = 'relative';
        this.el.style.zIndex = '1000';
    }

    updatePhysics(dt) {
        if (!this.isActive) return;

        // Kinematics: s = v0*t + 0.5*a*t^2; v = v0 + at
        let ds_x = this.velX * dt + 0.5 * this.accX * (dt * dt);
        let ds_y = this.velY * dt + 0.5 * this.accY * (dt * dt);

        this.x += ds_x;
        this.y += ds_y;

        this.velX += this.accX * dt;
        this.velY += this.accY * dt;

        this.angle += this.angularVel * dt;

        this.el.style.transform = `translate(${this.x}px, ${this.y}px) rotate(${this.angle}deg)`;
    }

    reset() {
        this.isActive = false;
        this.x = 0;
        this.y = 0;
        this.velX = 0;
        this.velY = 0;
        this.accX = 0;
        this.accY = 0;
        this.angle = 0;
        this.angularVel = 0;
        this.el.style.transform = `translate(0px, 0px) rotate(0deg)`;
        this.el.style.position = '';
        this.el.style.zIndex = '';
    }
}
